src/rebar_prv_install_deps.erl

```diff
@@ -184,33 +153,74 @@ handle_deps(Profile, State, Deps, Upgrade, Locks) ->
 %% Internal functions
 %% ===================================================================
 
+deps_per_profile(Profile, {Apps, State}) ->
+    Locks = rebar_state:get(State, {locks, Profile}, []),
+    ProfileDeps = rebar_state:get(State, {deps, Profile}, []),
+    {ok, NewApps, NewState} = handle_deps(Profile, State, ProfileDeps, Locks),
+    {NewApps++Apps, NewState}.
+
+find_cycles(Apps) ->
+    case rebar_digraph:compile_order(Apps) of
+        {error, {cycles, Cycles}} -> {cycles, Cycles};
+        {error, Error} -> {error, Error};
+        {ok, _} -> no_cycle
+    end.
+
+compile_order(Source, ProjectApps) ->
+    case rebar_digraph:compile_order(Source) of
+        {ok, Sort} ->
+            %% Valid apps are compiled and good
+            {ok, lists:dropwhile(fun rebar_app_info:valid/1, Sort -- ProjectApps)};
+        {error, Error} ->
+            {error, Error}
+    end.
+
+update_pkg_deps(Profile, Packages, PkgDeps, Graph, Upgrade, Seen, State) ->
+    case PkgDeps of
+        [] -> %% No pkg deps
+            {[], State};
+        PkgDeps ->
+            %% Find pkg deps needed
+            S = case rebar_digraph:cull_deps(Graph, PkgDeps) of
+                {ok, [], _} ->
+                    throw({rebar_digraph, no_solution});
+                {ok, Solution, []} ->
+                    Solution;
+                {ok, Solution, Discarded} ->
+                    [warn_skip_pkg(Pkg) || Pkg <- Discarded],
+                    Solution
+            end,
+            update_pkg_deps(Profile, S, Packages, Upgrade, Seen, State)
+    end.
+
 update_pkg_deps(Profile, Pkgs, Packages, Upgrade, Seen, State) ->
     %% Create app_info record for each pkg dep
     DepsDir = rebar_dir:deps_dir(State),
     {Solved, _, State1}
         = lists:foldl(fun(Pkg, {Acc, SeenAcc, StateAcc}) ->
-                              AppInfo = package_to_app(DepsDir
-                                                      ,Packages
-                                                      ,Pkg),
-                              {SeenAcc1, StateAcc1} = maybe_lock(Profile, AppInfo, SeenAcc, StateAcc, 0),
-                              case maybe_fetch(AppInfo, Upgrade, SeenAcc1, State) of
-                                  true ->
-                                      {[AppInfo | Acc], SeenAcc1, StateAcc1};
-                                  false ->
-                                      {Acc, SeenAcc1, StateAcc1}
-                              end
+                        handle_pkg_dep(Profile, Pkg, Packages, Upgrade, DepsDir, Acc, SeenAcc, StateAcc)
```

Comment:
What if instead of passing in all of these parameters you stored them in like a pkg_dep record and passed the record around instead?